contracts/hydro/src/contract.rs

```diff
@@ -2753,6 +2771,143 @@ pub fn convert_lockup_to_dtoken_reply(
+    let mut available_base_token: Uint128 = Uint128::zero();
+    let mut fund_conversion: Vec<(String, Uint128, Decimal)> = vec![]; // (denom, original amount, ratio)
+    for coin in &info.funds {
+        let token_group_id = token_manager
+            .validate_denom(&deps.as_ref(), current_round, coin.denom.clone())
+            .map_err(|err| {
+                new_generic_error(format!("Error validating denom '{}': {}", coin.denom, err))
+            })?;
+        let ratio = token_manager
+            .get_token_group_ratio(&deps.as_ref(), current_round, token_group_id)
+            .map_err(|err| {
+                new_generic_error(format!(
+                    "Error fetching token group ratio for '{}': {}",
+                    coin.denom, err
+                ))
+            })?;
+        let coin_as_base = Decimal::from_ratio(coin.amount, Uint128::one()).checked_mul(ratio)?;
+
+        available_base_token = available_base_token.checked_add(coin_as_base.to_uint_floor())?;
+
+        fund_conversion.push((coin.denom.clone(), coin.amount, ratio));
+    }
+
+    // Step 4: Determine how much to deduct and update state
+    // Check if payment is full or partial
+    if available_base_token >= slash_base_token.to_uint_floor() {
+        LOCKS_PENDING_SLASHES.remove(deps.storage, lockup.lock_id);
+    } else {
+        let remaining_base = slash_base_token
+            .checked_sub(Decimal::from_ratio(available_base_token, Uint128::one()))?;
+
+        let remaining_original = remaining_base.checked_div(slash_ratio)?;
+        LOCKS_PENDING_SLASHES.save(
+            deps.storage,
+            lockup.lock_id,
+            &remaining_original.to_uint_floor(),
+        )?;
+    }
+
+    // Step 5: Figure out what was used and what should be refunded
+    let mut funds_used: HashMap<String, Uint128> = HashMap::new();
+    let mut remaining_base_token =
+        core::cmp::min(slash_base_token.to_uint_floor(), available_base_token);
+
+    for (denom, original_amount, ratio) in &fund_conversion {
+        if remaining_base_token.is_zero() {
+            break;
+        }
+        let coin_as_base =
+            Decimal::from_ratio(*original_amount, Uint128::one()).checked_mul(*ratio)?;
+        let take_base = core::cmp::min(coin_as_base.to_uint_floor(), remaining_base_token);
+        let take_original = Decimal::from_ratio(take_base, Uint128::one()).checked_div(*ratio)?;
+
+        funds_used.insert(denom.clone(), take_original.to_uint_floor());
+        remaining_base_token = remaining_base_token.checked_sub(take_base)?
+    }
+
+    // Step 6: Prepare refund messages
+    let mut refund_msgs: Vec<BankMsg> = vec![];
+    for (denom, original_amount, _) in &fund_conversion {
+        let used = funds_used.get(denom).copied().unwrap_or(Uint128::zero());
+        let total = *original_amount;
+        if total > used {
+            let refund_amount = total.checked_sub(used).unwrap();
+            if refund_amount > Uint128::zero() {
```

Comment:
Use `?` instead of `.unwrap()`.

`if refund_amount > Uint128::zero() {` seems redundant, since we already checked `if total > used`, correct?